[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #84423 [8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations
* #83876 [6/N] [Dispatchable Collectives] Update recv with CPU / CUDA implementations
* #83859 [5/N] [Dispatchable Collectives] Update send with CPU / CUDA implementations
* #83810 [4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations
* **#83735 [3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations**

### About this PR
* Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Add test to validate that a separate device implementation is not supported.

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

H-Huang committed:
[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -147,8 +135,7 @@ TORCH_LIBRARY(c10d, m) {
   // __torch_dispatch__.
   m.def(
-      "broadcast_",
-      dispatch(c10::DispatchKey::CompositeExplicitAutograd, broadcast_));
```
Hey @bdhirsh, in the comment above, Jiewen mentioned "It's important to register the op to the CompositeExplicitAutograd key to enable __ torch_dispatch __.". 

With this new change we are not specifying a default implementation or the CompositeExplicitAutograd dispatch key, is this okay? Will __ torch_dispatch __ still work?

bdhirsh replied:
tldr is yep, it should still work fine.

(I think the thing Jiewen was referring to is that if you made your kernel `CompositeImplicitAutograd`, it wouldn't work with `__torch_dispatch__`. You're writing separate CPU + CUDA implementations though, which will work fine)

kwen2501 replied:
@H-Huang nit: shall we update the comment here given @bdhirsh's comment?

H-Huang replied:
Yep sounds good

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.


[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.


[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.


[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.


[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d_common.py:
```diff
@@ -1325,6 +1325,38 @@ def test_send_recv(self):
+        # correctly dispatched
+
+        # negative test to make sure a non-supported device fails during dispatch call
+        nonsupported_device = torch.device("meta")
```
maybe I'm missing something but it doesn't see like there's tests with CPU and GPU tensors?

H-Huang replied:
Right, since this is a big change we are trying to do it piecewise. As of this PR, ProcessGroup doesn't support a list of backends, so we are still using ProcessGroupNCCL and ProcessGroupGloo to test this. Later once ProcessGroup adds that support, this test will also be updated to test the dispatching of CPU and GPU tensors

kwen2501 commented on torch/csrc/distributed/c10d/OpsImpl.hpp:
```diff
@@ -0,0 +1,25 @@
+    int64_t timeout);
+
+} // namespace ops
+} // namespace c10d
```
nit: do we need this header file? 
Other than OpsImpl.cpp, anywhere else is the header file needed?

H-Huang replied:
Oh, that is a good point, I think we don't actually need it. Removing.

kwen2501 reviewed (approved):
LGTM! Just minor comments added. Thanks!

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### Changes
- Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
- Add test to validate that a separate device implementation is not supported.

#### Motivation
In the future PRs we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38876771).

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -6,21 +6,6 @@
 
 namespace c10d {
 namespace {
-std::tuple<std::vector<at::Tensor>, c10::intrusive_ptr<Work>> broadcast_(
```
I believe removing this default implementation will break custom Python PG, like the one in PyTorch/XLA.

alanwaketan replied:
The way how the Python PG works is that it relies on the underlying cc ops to call the PyProcessGroup.
Now that you remove the default implementation, and then the device registered is XLA. The PyProcessGroup will never be called.

H-Huang replied:
I will add it back, thanks

H-Huang committed:
Update on "[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations"


### About this PR
* Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Add test to validate that a separate device implementation is not supported.

### About this stack
In the future we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38876771).

H-Huang commented:
@pytorchbot merge

kwen2501 commented:
Nice job! This is so exciting!
